apps/music/js/ui/views/mode_manager.js

```diff
@@ -25,25 +38,136 @@ var ModeManager = {
     return this._modeStack[this._modeStack.length - 1];
   },
 
+  get previousMode() {
+    return this._modeStack[this._modeStack.length - 2];
+  },
+
   start: function(mode, callback) {
-    this._modeStack = [mode];
-    this._updateMode(callback);
+    this._modeStack = [];
+    this._resetViews();
+    this._pushView(this.views[mode], false, function() {
+      this._modeStack = [mode];
+      this._updateMode(callback);
+    }.bind(this));
   },
 
   push: function(mode, callback) {
-    this._modeStack.push(mode);
-    this._updateMode(callback);
+    this._pushView(this.views[mode], true, function() {
+      this._modeStack.push(mode);
+      this._updateMode(callback);
+    }.bind(this));
   },
 
   pop: function() {
     if (this._modeStack.length <= 1) {
       return;
     }
+    this._popView();
     this._modeStack.pop();
     this._updateMode();
   },
 
-  updateBackArrow: function() {
+  _resetViews: function() {
+    for (var mode in this.views) {
+      var sheet = document.getElementById(this.views[mode].id);
+      sheet.classList.remove('animated');
+      sheet.classList.remove('current');
+      sheet.classList.add('next');
+    }
+  },
+
+  initView: function(view, callback) {
+    if (!view.isLoaded) {
+      LazyLoader.load(view.path, function() {
+        // Remove the view's hidden style before pushing it.
+        var sheet = document.getElementById(view.id);
+        sheet.classList.remove('hidden');
+
+        switch(view.id) {
+          case 'views-tiles':
+            TilesView.init();
+            break;
+          case 'views-list':
+            ListView.init();
+            break;
+          case 'views-sublist':
+            SubListView.init();
+            break;
+          case 'views-player':
+            PlayerView.init();
+            break;
+          case 'views-search':
+            SearchView.init();
+            break;
+        }
+        view.isLoaded = true;
+
+        // The PlayerView needs the settings values before use it.
+        if (view.id === 'views-player') {
+          asyncStorage.getItem(SETTINGS_OPTION_KEY, function(settings) {
+            App.playerSettings = settings;
+            PlayerView.setOptions(App.playerSettings);
+            callback();
+          });
+        } else if (view.id === 'views-search') {
+          // The text normalizer is needed in search view.
+          LazyLoader.load('shared/js/text_normalizer.js', callback);
+        } else {
+          callback();
+        }
+      });
+    } else {
+      callback();
+    }
+  },
+
+  _pushView: function(view, animated, callback) {
+    this.initView(view, function() {
+      var nextId = view.id;
+      var currentId =
+        this.currentMode ? this.views[this.currentMode].id : null;
```

Comment:
Nit: Maybe this would be easier to read like so:

```
      var currentId = this.currentMode ? 
                      this.views[this.currentMode].id : null;
```